src/mvba/abba/mod.rs

```diff
@@ -1,56 +1,567 @@
+                                self.subject, subject
+                            )));
+                        }
+
+                        if !self
+                            .pub_key_set
+                            .public_key()
+                            .verify(proof, &subject.to_bytes())
+                        {
+                            return Err(Error::InvalidMessage("invalid proof".to_string()));
+                        }
+
+                        // A weaker validity: an honest party may only decide on a value
+                        // for which it has the accompanying validating data.
+                        if action.value != PreVoteValue::One {
+                            return Err(Error::InvalidMessage("invalid value".to_string()));
+                        }
+                    }
+                    PreVoteJustification::HardJustification(sig) => {
+                        // Soft pre-vote justification is the S-threshold signature for `(ID, pre-vote, r − 1, b)`
+                        let sign_bytes =
+                            self.pre_vote_bytes_to_sign(action.round - 1, &action.value)?;
+                        if !self.pub_key_set.public_key().verify(sig, &sign_bytes) {
+                            return Err(Error::InvalidMessage(
+                                "invalid soft-vote justification".to_string(),
+                            ));
+                        }
+                    }
+                    PreVoteJustification::SoftJustification(sig) => {
+                        // Hard pre-vote justification is the S-threshold signature for `(ID, main-vote, r − 1, abstain)`
+                        let sign_bytes =
+                            self.main_vote_bytes_to_sign(self.r - 1, &MainVoteValue::Abstain)?;
+                        if !self.pub_key_set.public_key().verify(sig, &sign_bytes) {
+                            return Err(Error::InvalidMessage(
+                                "invalid hard-vote justification".to_string(),
+                            ));
+                        }
+                    }
+                }
+            }
+            Action::MainVote(action) => {
+                // check the validity of the S-signature share
+                let sign_bytes = self.main_vote_bytes_to_sign(action.round, &action.value)?;
+                if !self
+                    .pub_key_set
+                    .public_key_share(sender)
+                    .verify(&action.sig_share, &sign_bytes)
+                {
+                    return Err(Error::InvalidMessage("invalid signature share".to_string()));
+                }
+
+                match &action.justification {
+                    MainVoteJustification::NoAbstainJustification(sig) => {
+                        // valid S-signature share on the message `(ID, pre-vote, r, b)`
+                        let sign_bytes =
+                            self.pre_vote_bytes_to_sign(action.round, &PreVoteValue::One)?;
```

Comment:
I think you want to pull out the value from the main vote. i.e. it could be a main vote for zero. This check would only accept main votes for one

Reply:
Maybe a party receives 3 pre-vote zeros. All n-t are zero.